generator/public_transport_processor.cpp

```diff
@@ -0,0 +1,160 @@
+#include "generator/public_transport_processor.hpp"
+
+#include "geometry/mercator.hpp"
+
+#include "base/assert.hpp"
+#include "base/logging.hpp"
+
+#include "defines.hpp"
+
+// Stops are nodes with tags:
+//  highway = bus_stop
+//  amenity = bus_stattion
+//  public_transport = stop_station
+
+// Routes are relations with tags:
+//  route = bus
+//  route = tram
+//  route = trolleybus
+// relation |ref| tag is a route name
+
+namespace
+{
+
+// Node keys
+string const kAmenity = "amenity";
+string const kHighway = "highway";
+string const kPublicTransport = "public_transport";
+
+// Node values
+string const kBusStop = "bus_stop";
+string const kBusStation = "bus_station";
+string const kStopPosition = "stop_position";
+
+// Relation keys
+string const kRef = "ref";
+string const kRoute = "route";
+
+// Relation values
+string const kBus = "bus";
+string const kTram = "tram";
+string const kTrolleybus = "trolleybus";
+
+}  // namespace
+
+bool PublicTransportProcessor::ProcessStop(OsmElement & node, FeatureParams & params)
+{
+  ASSERT_EQUAL(node.type, OsmElement::EntityType::Node, ());
+
+  if (node.GetTag(kPublicTransport) != kStopPosition &&
+      node.GetTag(kHighway) != kBusStop &&
+      node.GetTag(kAmenity) != kBusStation)
+    return false; // not a stop
+
+  StopInfo & si = m_stops[node.id];
+  si.params = params;
+  si.pt = MercatorBounds::FromLatLon(node.lat, node.lon);
+
+  auto i = m_routesInNode.find(node.id);
+  if (i != m_routesInNode.end())
+  {
+    si.routes.buses = move(i->second.buses);
+    si.routes.trams = move(i->second.trams);
+    si.routes.trolleybuses = move(i->second.trolleybuses);
+    m_routesInNode.erase(i);
+  }
+
+  return true;
+}
+
+bool PublicTransportProcessor::ProcessRoute(OsmElement & relation)
+{
+  ASSERT_EQUAL(relation.type, OsmElement::EntityType::Relation, ());
+
+  bool const isBus = relation.GetTag(kRoute) == kBus;
+  bool const isTram = !isBus && relation.GetTag(kRoute) == kTram;
+  bool const isTrolleybus = !isBus && !isTram && relation.GetTag(kRoute) == kTrolleybus;
+
+  if (!isBus && !isTram && !isTrolleybus)
+    return false; // not a route
+
+  string const ref = relation.GetTag(kRef);
+  if (ref.empty())
+    return false; // no route name, skip it
+
+  for (OsmElement::Member const & m : relation.Members())
+  {
+    if (m.type != OsmElement::EntityType::Node)
+      continue;
+
+    Routes * routes = nullptr;
+    auto const i = m_stops.find(m.ref);
+    if (i == m_stops.end())
+    {
+      routes = &(m_routesInNode[m.ref]);
```

Comment:
Why do we need such complex interaction between m_stops and m_routesInNode? Why not to just accumulate info in m_routesInNode and in m_stops, and then, at once, perform all merging logic in ForEachStop.

In other words, I thought that this dancing pointers and moving from m_routesInNode in ProcessStop are complex.